File: exercises/leetcode/algorithms/451-sort-characters-by-frequency/solution.ts
function frequencySort(s: string): string {
   // Build the frequency map
   const freqMap = {}
   for (let char of s) {
      freqMap[char] = (freqMap[char] || 0) + 1
   }

   // Create an array and sort it based on the frequency
   const sortedChars = Object.keys(freqMap).sort(
      (a, b) => freqMap[b] - freqMap[a]
   )

   // Construct the result string
   let result = ''
   for (let char of sortedChars) {
      result += char.repeat(freqMap[char])
   }

   return result
}

File: exercises/leetcode/algorithms/451-sort-characters-by-frequency/solution.py
class Solution:
    def frequencySort(self, s: str) -> str:
        # Build the frequency map
        freq_map = Counter(s)

        # Build the max heap
        max_heap = [(-freq, char) for char, freq in freq_map.items()]
        heapq.heapify(max_heap)

        # Extract from max heap and build result
        res = []
        while max_heap:
            freq, char = heapq.heappop(max_heap)
            res.append(char * (-freq))

        return ''.join(res)
